Crash when declaring a datatype const as the result of a function with reads

The following program causes Dafny to crash:

```
  class MyClass {
    ghost const repr : object
  }

  datatype D1 = D1(o : MyClass)
  {
    ghost const objs : set<object> := getObjs()

    function getObjs() : set<object>
      reads o
      { {o, o.repr} }
  }
```

The reads clause redundantly includes `o` (since `repr` is marked `const`). However rather than 
giving an error due to having a `reads` clause, Dafny crashes with the following message:

```
Dafny 2.3.0.10506
[ERROR] FATAL UNHANDLED EXCEPTION: System.Diagnostics.Contracts.ContractException: Assertion failed.
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailureImplementation (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0002f] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailure (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x00000] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Diagnostics.Contracts.Contract.ReportFailure (System.Diagnostics.Contracts.ContractFailureKind failureKind, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0003a] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Diagnostics.Contracts.Contract.Assert (System.Boolean condition) [0x00003] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at Microsoft.Boogie.NAryExpr..ctor (Microsoft.Boogie.IToken tok, Microsoft.Boogie.IAppliable fun, System.Collections.Generic.IList`1[T] args, System.Boolean immutable) [0x00043] in <9378f975b0154ba1afae7c7cf97e000c>:0
  at Microsoft.Dafny.Translator+ExpressionTranslator.TrExpr (Microsoft.Dafny.Expression expr) [0x017d1] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator+ExpressionTranslator.TrExpr (Microsoft.Dafny.Expression expr) [0x055f4] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.GetReadonlyField (Microsoft.Dafny.Field f) [0x00452] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.AddClassMembers (Microsoft.Dafny.TopLevelDeclWithMembers c, System.Boolean includeMethods) [0x002c0] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.AddTypeDecl (Microsoft.Dafny.RevealableTypeDecl d) [0x000ca] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.DoTranslation (Microsoft.Dafny.Program p, Microsoft.Dafny.ModuleDefinition forModule) [0x00437] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator+<Translate>d__47.MoveNext () [0x000ea] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.DafnyDriver+<Translate>d__6.MoveNext () [0x00119] in <cd1f2aa755c64f5bbfa0950a1869826c>:0
  at Microsoft.Dafny.DafnyDriver.Boogie (System.String baseName, System.Collections.Generic.IEnumerable`1[T] boogiePrograms, System.String programId, System.Collections.Generic.Dictionary`2[System.String,Microsoft.Boogie.PipelineStatistics]& statss, Microsoft.Boogie.PipelineOutcome& oc) [0x0012f] in <cd1f2aa755c64f5bbfa0950a1869826c>:0
  at Microsoft.Dafny.DafnyDriver.ProcessFiles (System.Collections.Generic.IList`1[T] dafnyFiles, System.Collections.ObjectModel.ReadOnlyCollection`1[T] otherFileNames, Microsoft.Dafny.ErrorReporter reporter, System.Boolean lookForSnapshots, System.String programId) [0x00222] in <cd1f2aa755c64f5bbfa0950a1869826c>:0
  at Microsoft.Dafny.DafnyDriver.ThreadMain (System.String[] args) [0x0003a] in <cd1f2aa755c64f5bbfa0950a1869826c>:0
  at Microsoft.Dafny.DafnyDriver+<>c__DisplayClass1_0.<Main>b__0 () [0x00001] in <cd1f2aa755c64f5bbfa0950a1869826c>:0
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) [0x00014] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x0002b] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Threading.ThreadHelper.ThreadStart () [0x00008] in <f9d1b832704f410aa8ec771f4fe80552>:0
```

--- a/Source/Dafny/Translator.cs
+++ b/Source/Dafny/Translator.cs
@@ -1282,7 +1282,7 @@ namespace Microsoft.Dafny {
         name += "$Is";
         // $Is(o, ..)
         is_o = MkIs(o, o_ty, ModeledAsBoxType(dd.Var.Type));
-        var etran = new ExpressionTranslator(this, predef, new Bpl.IdentifierExpr(dd.tok, "$OneHeap", predef.HeapType));
+        var etran = new ExpressionTranslator(this, predef, NewOneHeapExpr(dd.tok));
         Bpl.Expr parentConstraint, constraint;
         if (dd.Var.Type.IsNumericBased() || dd.Var.Type.IsBitVectorType || dd.Var.Type.IsBoolType) {
           // optimize this to only use the numeric/bitvector constraint, not the whole $Is thing on the base type
@@ -8486,7 +8486,7 @@ namespace Microsoft.Dafny {
       return name;
     }
 
-    private Expr newOneHeapExpr(IToken tok) {
+    private Expr NewOneHeapExpr(IToken tok) {
       return new Bpl.IdentifierExpr(tok, "$OneHeap", predef.HeapType);
     }
 
@@ -8703,7 +8703,7 @@ namespace Microsoft.Dafny {
         {
           var bvars = new List<Bpl.Variable>();
           var types = Map(Enumerable.Range(0, arity + 1), i => BplBoundVar("t" + i, predef.Ty, bvars));
-          var oneheap = newOneHeapExpr(tok);
+          var oneheap = NewOneHeapExpr(tok);
           var h = BplBoundVar("heap", predef.HeapType, bvars);
           var f = BplBoundVar("f", predef.HandleType, bvars);
           var boxes = Map(Enumerable.Range(0, arity), i => BplBoundVar("bx" + i, predef.BoxType, bvars));
@@ -8743,7 +8743,7 @@ namespace Microsoft.Dafny {
         {
           var bvars = new List<Bpl.Variable>();
           var types = Map(Enumerable.Range(0, arity + 1), i => BplBoundVar("t" + i, predef.Ty, bvars));
-          var oneheap = newOneHeapExpr(tok);
+          var oneheap = NewOneHeapExpr(tok);
           var h = BplBoundVar("heap", predef.HeapType, bvars);
           var f = BplBoundVar("f", predef.HandleType, bvars);
           var boxes = Map(Enumerable.Range(0, arity), i => BplBoundVar("bx" + i, predef.BoxType, bvars));
@@ -9156,7 +9156,7 @@ namespace Microsoft.Dafny {
           // function QQ():int { 3 }
           var cf = (ConstantField)f;
           if (cf.Rhs != null && RevealedInScope(cf)) {
-            var etran = new ExpressionTranslator(this, predef, newOneHeapExpr(f.tok));
+            var etran = new ExpressionTranslator(this, predef, NewOneHeapExpr(f.tok));
             ff.Body = etran.TrExpr(cf.Rhs);
           }
           sink.AddTopLevelDeclaration(ff);

--- a/Source/Dafny/Translator.cs
+++ b/Source/Dafny/Translator.cs
@@ -8486,6 +8486,10 @@ namespace Microsoft.Dafny {
       return name;
     }
 
+    private Expr newOneHeapExpr(IToken tok) {
+      return new Bpl.IdentifierExpr(tok, "$OneHeap", predef.HeapType);
+    }
+
     private void AddArrowTypeAxioms(ArrowTypeDecl ad) {
       Contract.Requires(ad != null);
       var arity = ad.Arity;
@@ -8699,7 +8703,7 @@ namespace Microsoft.Dafny {
         {
           var bvars = new List<Bpl.Variable>();
           var types = Map(Enumerable.Range(0, arity + 1), i => BplBoundVar("t" + i, predef.Ty, bvars));
-          var oneheap = new Bpl.IdentifierExpr(tok, "$OneHeap", predef.HeapType);
+          var oneheap = newOneHeapExpr(tok);
           var h = BplBoundVar("heap", predef.HeapType, bvars);
           var f = BplBoundVar("f", predef.HandleType, bvars);
           var boxes = Map(Enumerable.Range(0, arity), i => BplBoundVar("bx" + i, predef.BoxType, bvars));
@@ -8739,7 +8743,7 @@ namespace Microsoft.Dafny {
         {
           var bvars = new List<Bpl.Variable>();
           var types = Map(Enumerable.Range(0, arity + 1), i => BplBoundVar("t" + i, predef.Ty, bvars));
-          var oneheap = new Bpl.IdentifierExpr(tok, "$OneHeap", predef.HeapType);
+          var oneheap = newOneHeapExpr(tok);
           var h = BplBoundVar("heap", predef.HeapType, bvars);
           var f = BplBoundVar("f", predef.HandleType, bvars);
           var boxes = Map(Enumerable.Range(0, arity), i => BplBoundVar("bx" + i, predef.BoxType, bvars));
@@ -9152,7 +9156,7 @@ namespace Microsoft.Dafny {
           // function QQ():int { 3 }
           var cf = (ConstantField)f;
           if (cf.Rhs != null && RevealedInScope(cf)) {
-            var etran = new ExpressionTranslator(this, predef, (Bpl.Expr)null);
+            var etran = new ExpressionTranslator(this, predef, newOneHeapExpr(f.tok));
             ff.Body = etran.TrExpr(cf.Rhs);
           }
           sink.AddTopLevelDeclaration(ff);

--- a/Test/git-issues/git-issue-615.dfy
+++ b/Test/git-issues/git-issue-615.dfy
@@ -1,12 +1,26 @@
-  class MyClass {
-    ghost const repr: object
-  }
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
 
-  datatype D1 = D1(o: MyClass)
+class MyClass {
+  ghost const repr: object
+}
+
+datatype D1 = D1(o: MyClass)
+{
+  ghost const objs: set<object> := getObjs()
+
+  function getObjs(): set<object>
+    reads o
+    { {o, o.repr} }
+}
+
+class MyClass2 {
+  var x: int
+  const c := F()
+
+  function method F(): int
+    reads this
   {
-    ghost const objs: set<object> := getObjs()
-
-    function getObjs(): set<object>
-      reads o
-      { {o, o.repr} }
+    x
   }
+}

--- a/Test/git-issues/git-issue-615.dfy
+++ b/Test/git-issues/git-issue-615.dfy
@@ -1,0 +1,12 @@
+  class MyClass {
+    ghost const repr : object
+  }
+
+  datatype D1 = D1(o : MyClass)
+  {
+    ghost const objs : set<object> := getObjs()
+
+    function getObjs() : set<object>
+      reads o
+      { {o, o.repr} }
+  }

--- a/Test/git-issues/git-issue-615.dfy
+++ b/Test/git-issues/git-issue-615.dfy
@@ -1,12 +1,12 @@
   class MyClass {
-    ghost const repr : object
+    ghost const repr: object
   }
 
-  datatype D1 = D1(o : MyClass)
+  datatype D1 = D1(o: MyClass)
   {
-    ghost const objs : set<object> := getObjs()
+    ghost const objs: set<object> := getObjs()
 
-    function getObjs() : set<object>
+    function getObjs(): set<object>
       reads o
       { {o, o.repr} }
   }

--- a/Test/git-issues/git-issue-615.dfy.expect
+++ b/Test/git-issues/git-issue-615.dfy.expect
@@ -1,0 +1,8 @@
+git-issue-615.dfy(10,35): Error: insufficient reads clause to invoke function
+Execution trace:
+    (0,0): anon0
+git-issue-615.dfy(19,13): Error: insufficient reads clause to invoke function
+Execution trace:
+    (0,0): anon0
+
+Dafny program verifier finished with 2 verified, 2 errors

